File: Arrays/88. Merge Sorted Array.py
from typing import List

#brute force O((m+n)log(m+n)) + O(m)
# class Solution:
#     def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
#         for i in range(len(nums2)):
#             nums1[m+i] = nums2[i]
#         nums1.sort()

#optimal O(m+n)
class Solution: 
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = m - 1
        j = n - 1
        k = len(nums1) - 1
        while i >= 0 and j >= 0:
            if nums2[j] > nums1[i]:
                nums1[k] = nums2[j]
                j -= 1
            else:
                nums1[k] = nums1[i]
                i -=1
            k -= 1
        while j >= 0:
            nums1[k] = nums2[j]
            j -= 1
            k -= 1
        
        
if __name__ == "__main__":
    sol = Solution()
    nums1 = [1,2,3,0,0,0]
    m = 3
    nums2 = [2,5,6]
    n = 3
    sol.merge(nums1, m, nums2, n)
    print(nums1)

File: Arrays & Strings/88. Merge Sorted Array.cpp
/*
iterate through the nums1 in reverse order
compare elements of nums1 from m to 0 with the elements of nums2
add the largest num to the current position of the nums1
*/

#include <bits/stdc++.h>

class Solution {
public:
    void merge(std::vector<int>& nums1, int m, std::vector<int>& nums2, int n) {
        int j = m - 1, k = n - 1;

        for (int i = nums1.size() - 1; i >= 0 && k >= 0; i--) {
            if (j >= 0 && nums1[j] > nums2[k]) {
                nums1[i] = nums1[j--];
            }
            else {
                nums1[i] = nums2[k--];
            }
        }      

        for (auto num: nums1) {
            std::cout << num;
        }

        std::cout << std::endl;
    }
};


int main() {
    std::vector<int> nums1 = {0};
    int m = 0;
    std::vector<int> nums2 = {1};
    int n = 1;
    
    Solution sol;
    sol.merge(nums1, m, nums2, n);        
    
    return 0;
}
